File: mangledotdev-v0.1.0_Python/mangledotdev.py
import subprocess
import json
import time
import random
import os
import sys
import io
import uuid


class InputManager:
    """
    Manages sending requests to other processes and handling responses.
    
    This is an instance-based class - create one instance per request.
    
    Attributes:
        response (dict): Complete response with status, data, errors, warnings
    
    Methods:
        request(): Send a request to another process
        get_response(): Get the full response object
        get_data(): Get the response data (returns None on error)
    """
    
    @staticmethod
    def __genKey():
        """Generate a unique key for request/response matching."""
        return uuid.uuid4().hex

    def __init__(self):
        """Initialize a new InputManager instance."""
        self.__process = None
        self.__raw_request = None
        self.__request = None
        self.__response = None
        self.response = None
        self.__data = None
        self.__key = None

    def __get_command(self, language, file):
        """
        Validate file and build command to execute.

        Args:
            language (str): Programming language/runtime
            file (str): Path to file to execute

        Returns:
            list: Command array for subprocess

        Raises:
            ValueError: Invalid file extension for language
            FileNotFoundError: File does not exist
            PermissionError: File not readable/executable
        """
        lang_upper = str(language).upper()

        # On Windows, convert forward slashes to backslashes for file system operations
        if os.name == 'nt':
            file = file.replace('/', '\\')

        file_ext = os.path.splitext(file)[1].lower()

        # Extension validation - FIRST before file existence check
        extension_map = {
            'PYTHON': ['.py'],
            'PY': ['.py'],
            'JAVASCRIPT': ['.js'],
            'JS': ['.js'],
            'NODE': ['.js'],
            'NODEJS': ['.js'],
            'RUBY': ['.rb'],
            'RB': ['.rb'],
            'C': ['.c', '.out', '.exe', ''],
            'CS': ['.exe', '.dll', ''],
            'CPP': ['.cpp', '.cc', '.cxx', '.out', '.exe', ''],
            'C#': ['.exe', '.dll', ''],
            'C++': ['.cpp', '.cc', '.cxx', '.out', '.exe', ''],
            'CSHARP': ['.exe', '.dll', ''],
            'CPLUSPLUS': ['.cpp', '.cc', '.cxx', '.out', '.exe', ''],
            'EXE': ['.cpp', '.cc', '.cxx', '.out', '.exe', ''],
            'JAR': ['.jar'],
            'JAVA': ['.jar'],
            'RUST': ['.rs', '.exe', '.out', ''],
            'RS': ['.rs', '.exe', '.out', ''],
            'GO': ['.go', '.exe', '.out', ''],
            'GOLANG': ['.go', '.exe', '.out', '']
        }

        if lang_upper in extension_map:
            valid_extensions = extension_map[lang_upper]
            if file_ext not in valid_extensions:
                expected = ', '.join([ext if ext else '(no extension)' for ext in valid_extensions])
                raise ValueError(f"Invalid file '{file}' for language '{language}'. Expected: e.g. 'file{expected}'")

        # File existence check
        if not os.path.exists(file):
            raise FileNotFoundError(f"File not found: {file}")

        if not os.path.isfile(file):
            raise ValueError(f"Path is not a file: {file}")

        # Permission checks
        compiled_langs = ['C', 'CS', 'CPP', 'C#', 'C++', 'CSHARP', 'CPLUSPLUS', 'EXE', 'RUST', 'RS', 'GO', 'GOLANG']
        if lang_upper in compiled_langs:
            if not os.access(file, os.X_OK):
                raise PermissionError(f"File is not executable: {file}")
        else:
            if not os.access(file, os.R_OK):
                raise PermissionError(f"File is not readable: {file}")

        # Auto-add ./ for compiled executables if not present and not absolute path
        if lang_upper in compiled_langs and not os.path.isabs(file) and not file.startswith('./'):
            file = './' + file

        # Build command
        lang_map = {
            'PYTHON': ['python', file],
            'PY': ['python', file],
            'JAVASCRIPT': ['node', file],
            'JS': ['node', file],
            'NODE': ['node', file],
            'NODEJS': ['node', file],
            'RUBY': ['ruby', file],
            'RB': ['ruby', file],
            'C': [file],
            'CS': ['dotnet', file] if file_ext == '.dll' else [file],
            'CPP': [file],
            'C#': ['dotnet', file] if file_ext == '.dll' else [file],
            'C++': [file],
            'CSHARP': ['dotnet', file] if file_ext == '.dll' else [file],
            'CPLUSPLUS': [file],
            'EXE': [file],
            'JAR': ['java', '-jar', file],
            'JAVA': ['java', '-jar', file],
            'RUST': [file],
            'RS': [file],
            'GO': ['go', 'run', file] if file_ext == '.go' else [file],
            'GOLANG': ['go', 'run', file] if file_ext == '.go' else [file]
        }

        if lang_upper not in lang_map:
            raise ValueError(f"Unsupported language: {language}")

        return lang_map[lang_upper]

    def request(self, isUnique=True, optionalOutput=True, data: any=None, language: any=str, file: any=str):
        """
        Send a request to another process.
        
        Args:
            isUnique (bool): Expect single output (True) or multiple (False)
            optionalOutput (bool): Output is optional (True) or required (False)
            data: Data to send (any JSON-serializable type)
            language (str): Target language/runtime
            file (str): Path to target file
            
        Sets:
            self.response (dict): Complete response with keys:
                - request_status (bool|None): Success status
                - data: Response data
                - optionalOutput (bool): Echo of parameter
                - isUnique (bool): Echo of parameter
                - warnings (list): Warning messages
                - errors (list): Error messages
        """
        try:
            self.__key = InputManager._InputManager__genKey()
            command = self.__get_command(language, file)

            self.__process = subprocess.Popen(
                command,
                stdin=subprocess.PIPE,
                stdout=subprocess.PIPE,
                stderr=subprocess.PIPE,
                text=True
            )

            self.__raw_request = {
                "key": self.__key,
                "optionalOutput": optionalOutput,
                "isUnique": isUnique,
                "data": data
            }
            self.__request = json.dumps(self.__raw_request)

            response = {
                "request_status": None,
                "data": None,
                "optionalOutput": optionalOutput,
                "isUnique": isUnique,
                "warnings": [],
                "errors": []
            }

            output, errors = self.__process.communicate(input=self.__request)

            if self.__process.returncode != 0:
                response["request_status"] = False
                response["errors"].append(f"Process exited with code {self.__process.returncode}")
                if errors.strip():
                    response["errors"].append(f"stderr: {errors.strip()}")
                response["warnings"].append("Warning: these kind of errors result from an error in the targeted script.")
                self.response = response
                return

            self.__response = []
            for line in output.strip().split('\n'):
                if not line.strip():
                    continue
                
                try:
                    __data = json.loads(line)
                    
                    # Validate response has matching key or null key (for init errors)
                    # This ensures we only process responses meant for this request
                    if isinstance(__data, dict) and (__data.get('key') == self.__key or __data.get('key') is None):
                        self.__response.append(__data)
                except json.JSONDecodeError:
                    # Ignore lines that aren't valid JSON (e.g., debug prints)
                    pass

            if len(self.__response) != 0:
                failure = False
                for i in self.__response:
                    if not i["request_status"]:
                        failure = True

                if not failure:
                    response["request_status"] = True
                else:
                    response["request_status"] = False

                response["isUnique"] = self.__response[0]["isUnique"]

                for i in self.__response:
                    for index, e in enumerate(i["errors"]):
                        response["errors"].append(i["errors"][index])
                
                data_list = []
                for i in self.__response:
                    data_list.append(i["data"])
                
                if response["isUnique"]:
                    if len(data_list) == 1:
                        response["data"] = data_list[0]
                    else:
                        response["request_status"] = False
                        response["data"] = None
                        response["errors"].append(f"Error: Expected 1 output (isUnique=True) but received {len(data_list)}.")
                else:
                    response["data"] = data_list
                    
            elif optionalOutput:
                response["request_status"] = None
                response["data"] = None
                response["warnings"].append("Warning: the output setting is set to optional, and the targeted program didn't gave any output.")
            else:
                response["request_status"] = False
                response["data"] = None
                response["errors"].append("Error: OutputManager might not be used or not correctly.")

            self.response = response

        except (FileNotFoundError, PermissionError, ValueError) as e:
            self.response = {
                "request_status": False,
                "data": None,
                "optionalOutput": optionalOutput,
                "isUnique": isUnique,
                "warnings": ["Warning: targeted file not found or can't be executed, consider checking file informations and language dependencies."],
                "errors": [f"Error: {str(e)}"]
            }
        except Exception as e:
            self.response = {
                "request_status": False,
                "data": None,
                "optionalOutput": optionalOutput,
                "isUnique": isUnique,
                "warnings": [],
                "errors": [f"Unexpected error: {str(e)}"]
            }
        return

    def get_response(self):
        """
        Get the full response object.

        Returns:
            dict: The complete response with status, data, errors, warnings.
                  Keys: request_status, data, optionalOutput, isUnique, warnings, errors
        """
        return self.response

    def get_data(self):
        """
        Get the response data if request was successful.

        Returns:
            The data from the response (any type), or None if request failed.
            The return type matches the type sent by the target process.
        """
        if self.response:
            if self.response["request_status"]:
                return self.response["data"]
        return None


class OutputManager:
    """
    Manages receiving requests from other processes and sending responses.
    
    This is a class-based/static manager - all methods are class methods.
    Must call init() before using.
    
    Class Attributes:
        data: The request data (accessible after init())
    
    Class Methods:
        init(): Initialize and read request from stdin
        output(val): Send response back via stdout
    """
    
    __original_stdout = None
    __request = None
    __data = None
    data = None
    __request_status = None
    __optional = None
    __unique_state = None
    __init_error = None
    __errors = []
    __warnings = []

    @classmethod
    def init(cls):
        """
        Initialize OutputManager and read request from stdin.
        
        Must be called before using output() or accessing data.
        Suppresses stdout to prevent pollution of JSON protocol.
        """
        # Save original stdout so we can restore it later
        cls.__original_stdout = sys.stdout
        # Redirect stdout to StringIO to suppress all print statements
        sys.stdout = io.StringIO()
        # Read the entire stdin (the JSON request from InputManager)
        cls.__request = sys.stdin.read()
        cls.__data = json.loads(cls.__request)
        cls.data = cls.__data["data"]
        cls.__optional = cls.__data["optionalOutput"]
        
        # Reset state for new request
        cls.__errors = []
        cls.__warnings = []
        cls.__init_error = None
        cls.__request_status = None
        cls.__unique_state = None

    @classmethod
    def __write(cls, args, _data):
        """
        Internal method to write JSON response to stdout.
        
        Args:
            args: Data to send in response
            _data: Request data for key/metadata
        """
        # Restore original stdout to actually write the response
        if cls.__original_stdout:
            sys.stdout = cls.__original_stdout
        else:
            sys.stdout = sys.__stdout__
        
        # Build and write JSON response
        response = json.dumps({
            "key": _data["key"] if _data else None,
            "request_status": cls.__request_status, 
            "data": args,
            "optionalOutput": cls.__optional,
            "isUnique": _data["isUnique"] if _data else None,
            "errors": cls.__errors, 
            "warnings": cls.__warnings
        })
        
        sys.stdout.write(response)
        sys.stdout.write("\n")
        sys.stdout.flush()

    @classmethod
    def output(cls, val):
        """
        Send response back to the calling process.
        
        Args:
            val: Data to send (any JSON-serializable type)
            
        Note:
            Can be called multiple times if isUnique=False in request.
            Will error if called multiple times when isUnique=True.
        """
        # Check if OutputManager was initialized
        if not cls.__data:
            if not cls.__init_error:
                if cls.__original_stdout:
                    sys.stdout = cls.__original_stdout
                else:
                    sys.stdout = sys.__stdout__

                cls.__request_status = False
                cls.__errors.append("Error: OutputManager isn't initialized.")
                cls._OutputManager__write(args=None, _data=cls.__data)
                cls.__init_error = True

        else:
            # Check if we can output based on isUnique setting
            # unique_state tracks if we've already output once
            if not cls.__unique_state or not cls.__data["isUnique"]:
                cls.__request_status = True
                cls._OutputManager__write(args=val, _data=cls.__data)
            else:
                # Multiple outputs when isUnique=True is an error
                cls.__request_status = False
                cls.__errors.append(f"Error: outputs out of bound (isUnique: {cls.__unique_state}).")
                cls._OutputManager__write(args=val, _data=cls.__data)

            # Mark that we've output once
            cls.__unique_state = cls.__data["isUnique"]
            # Re-suppress stdout after writing response
            sys.stdout = io.StringIO()

File: C#/Mangledotdev.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;
using System.Linq;
using System.Security.Cryptography;
using System.Text.Json;
using System.Text.Json.Nodes;

namespace Mangledotdev
{
    /// <summary>
    /// InputManagerResponse - Contains the complete response structure
    ///
    /// Fields:
    ///     RequestStatusSet: Whether request status has been set
    ///     RequestStatus: Success status of the request
    ///     Data: Response data as JSON string (preserves type)
    ///     OptionalOutput: Echo of request parameter
    ///     IsUnique: Echo of request parameter
    ///     Warnings: List of warning messages
    ///     Errors: List of error messages
    /// </summary>
    public class InputManagerResponse
    {
        public bool RequestStatusSet { get; set; }
        public bool RequestStatus { get; set; }
        public string Data { get; set; }
        public bool OptionalOutput { get; set; }
        public bool IsUnique { get; set; }
        public List<string> Warnings { get; set; }
        public List<string> Errors { get; set; }

        public InputManagerResponse()
        {
            RequestStatusSet = false;
            RequestStatus = false;
            Data = "";
            OptionalOutput = true;
            IsUnique = true;
            Warnings = new List<string>();
            Errors = new List<string>();
        }
    }

    /// <summary>
    /// InputManager - Manages sending requests to other processes and handling responses
    ///
    /// This is an instance-based class - create one instance per request.
    ///
    /// Properties:
    ///     Response: Complete response with status, data, errors, warnings
    ///
    /// Methods:
    ///     Request(): Send a request to another process
    ///     GetResponse(): Get the full response object
    ///     GetData(): Get the response data (returns empty string on error)
    ///     Bundle(): Convert any value to JSON string
    /// </summary>
    public class InputManager
    {
        private string _key;
        private string _rawRequest;
        private string _request;
        private List<JsonNode> _responseData;

        public InputManagerResponse Response { get; set; }

        /// <summary>
        /// Create a new InputManager instance
        /// </summary>
        public InputManager()
        {
            _key = "";
            _rawRequest = "";
            _request = "";
            _responseData = new List<JsonNode>();
            Response = new InputManagerResponse();
        }

        /// <summary>
        /// Generate a unique key for request/response matching
        /// </summary>
        /// <returns>Unique key as 32-char hex string</returns>
        private static string GenKey()
        {
            byte[] bytes = RandomNumberGenerator.GetBytes(16);
            return Convert.ToHexString(bytes).ToLower();
        }

        /// <summary>
        /// Bundle any value to JSON string for use with Request()
        /// </summary>
        /// <param name="value">Any JSON-serializable value</param>
        /// <returns>JSON string representation</returns>
        public static string Bundle(object value)
        {
            return JsonSerializer.Serialize(value);
        }

        /// <summary>
        /// Validate file and build command to execute
        /// </summary>
        /// <param name="language">Programming language/runtime</param>
        /// <param name="file">Path to file to execute</param>
        /// <returns>Command string</returns>
        /// <exception cref="ArgumentException">Invalid file extension or file not found</exception>
        private string GetCommand(string language, string file)
        {
            string langUpper = language.ToUpper();

            // Normalize path separators for the current platform
            string filePath = file.Replace('/', Path.DirectorySeparatorChar);

            string ext = Path.GetExtension(filePath).ToLower();

            // Extension validation - FIRST before file existence check
            var extensionMap = new Dictionary<string, List<string>>
            {
                {"PYTHON", new List<string> {".py"}},
                {"PY", new List<string> {".py"}},
                {"JAVASCRIPT", new List<string> {".js"}},
                {"JS", new List<string> {".js"}},
                {"NODE", new List<string> {".js"}},
                {"NODEJS", new List<string> {".js"}},
                {"RUBY", new List<string> {".rb"}},
                {"RB", new List<string> {".rb"}},
                {"C", new List<string> {".c", ".out", ".exe", ""}},
                {"CS", new List<string> {".dll", ".exe", ""}},
                {"CPP", new List<string> {".cpp", ".cc", ".cxx", ".out", ".exe", ""}},
                {"C#", new List<string> {".dll", ".exe", ""}},
                {"C++", new List<string> {".cpp", ".cc", ".cxx", ".out", ".exe", ""}},
                {"CSHARP", new List<string> {".dll", ".exe", ""}},
                {"CPLUSPLUS", new List<string> {".cpp", ".cc", ".cxx", ".out", ".exe", ""}},
                {"EXE", new List<string> {".cpp", ".cc", ".cxx", ".out", ".exe", ""}},
                {"JAR", new List<string> {".jar"}},
                {"JAVA", new List<string> {".jar", ".class"}},
                {"RUST", new List<string> {".rs", ".exe", ".out", ""}},
                {"RS", new List<string> {".rs", ".exe", ".out", ""}},
                {"GO", new List<string> {".go", ".exe", ".out", ""}},
                {"GOLANG", new List<string> {".go", ".exe", ".out", ""}}
            };

            if (extensionMap.ContainsKey(langUpper))
            {
                var validExts = extensionMap[langUpper];
                if (!validExts.Contains(ext))
                {
                    string expected = string.Join(", ", validExts.Select(e => string.IsNullOrEmpty(e) ? "(no extension)" : e));
                    throw new ArgumentException($"Invalid file '{filePath}' for language '{language}'. Expected: e.g. 'file{expected}'");
                }
            }

            // File existence check
            if (!File.Exists(filePath))
            {
                throw new FileNotFoundException($"File not found: {filePath}");
            }

            // Permission checks
            var compiledLangs = new List<string> {"C", "CS", "CPP", "C#", "C++", "CSHARP", "CPLUSPLUS", "EXE", "RUST", "RS", "GO", "GOLANG"};
            // Note: C# doesn't have a reliable way to check execute permissions on Windows

            // Auto-add .\ or ./ for compiled executables if not present and not absolute path
            string finalPath = filePath;
            if (compiledLangs.Contains(langUpper) && !Path.IsPathRooted(filePath) &&
                !filePath.StartsWith("./") && !filePath.StartsWith(".\\") &&
                !filePath.StartsWith("." + Path.DirectorySeparatorChar))
            {
                finalPath = "." + Path.DirectorySeparatorChar + filePath;
            }

            // Build command
            var langMap = new Dictionary<string, string>
            {
                {"PYTHON", $"python {finalPath}"},
                {"PY", $"python {finalPath}"},
                {"JAVASCRIPT", $"node {finalPath}"},
                {"JS", $"node {finalPath}"},
                {"NODE", $"node {finalPath}"},
                {"NODEJS", $"node {finalPath}"},
                {"RUBY", $"ruby {finalPath}"},
                {"RB", $"ruby {finalPath}"},
                {"C", finalPath},
                {"CS", ext == ".dll" ? $"dotnet {finalPath}" : finalPath},
                {"CPP", finalPath},
                {"C#", ext == ".dll" ? $"dotnet {finalPath}" : finalPath},
                {"C++", finalPath},
                {"CSHARP", ext == ".dll" ? $"dotnet {finalPath}" : finalPath},
                {"CPLUSPLUS", finalPath},
                {"EXE", finalPath},
                {"JAR", $"java -jar {finalPath}"},
                {"JAVA", ext == ".jar" ? $"java -jar {finalPath}" : $"java -cp {Path.GetDirectoryName(finalPath) ?? "."} {Path.GetFileNameWithoutExtension(finalPath)}"},
                {"RUST", finalPath},
                {"RS", finalPath},
                {"GO", ext == ".go" ? $"go run {finalPath}" : finalPath},
                {"GOLANG", ext == ".go" ? $"go run {finalPath}" : finalPath}
            };

            if (!langMap.ContainsKey(langUpper))
            {
                throw new ArgumentException($"Unsupported language: {language}");
            }

            return langMap[langUpper];
        }

        /// <summary>
        /// Send a request to another process
        /// </summary>
        /// <param name="isUnique">Expect single output (true) or multiple (false)</param>
        /// <param name="optionalOutput">Output is optional (true) or required (false)</param>
        /// <param name="data">Data to send as JSON string</param>
        /// <param name="language">Target language/runtime</param>
        /// <param name="file">Path to target file</param>
        ///
        /// Sets Response property with fields:
        ///     - RequestStatus (bool): Success status
        ///     - Data (string): Response data as JSON string (preserves type)
        ///     - OptionalOutput (bool): Echo of parameter
        ///     - IsUnique (bool): Echo of parameter
        ///     - Warnings (List<string>): Warning messages
        ///     - Errors (List<string>): Error messages
        public void Request(bool isUnique = true, bool optionalOutput = true,
                          string data = "", string language = "", string file = "")
        {
            try
            {
                _key = GenKey();
                string command = GetCommand(language, file);

                // Clear previous response
                Response = new InputManagerResponse
                {
                    OptionalOutput = optionalOutput,
                    IsUnique = isUnique
                };

                // Build request JSON
                var requestObj = new JsonObject
                {
                    ["key"] = _key,
                    ["optionalOutput"] = optionalOutput,
                    ["isUnique"] = isUnique
                };

                if (!string.IsNullOrEmpty(data))
                {
                    try
                    {
                        requestObj["data"] = JsonNode.Parse(data);
                    }
                    catch
                    {
                        requestObj["data"] = null;
                    }
                }
                else
                {
                    requestObj["data"] = null;
                }

                string requestStr = requestObj.ToJsonString();

                // Parse command into executable and arguments
                string[] cmdParts = command.Split(' ', 2);
                string executable = cmdParts[0];
                string arguments = cmdParts.Length > 1 ? cmdParts[1] : "";

                ProcessStartInfo psi = new ProcessStartInfo
                {
                    FileName = executable,
                    Arguments = arguments,
                    UseShellExecute = false,
                    CreateNoWindow = true,
                    RedirectStandardInput = true,
                    RedirectStandardOutput = true,
                    RedirectStandardError = true
                };

                using (Process process = new Process { StartInfo = psi })
                {
                    process.Start();

                    // Write request to stdin
                    process.StandardInput.Write(requestStr);
                    process.StandardInput.Close();

                    // Read output from stdout
                    string output = process.StandardOutput.ReadToEnd();
                    string errors = process.StandardError.ReadToEnd();

                    process.WaitForExit();

                    // Check exit code
                    if (process.ExitCode != 0)
                    {
                        Response.RequestStatus = false;
                        Response.RequestStatusSet = true;
                        Response.Errors.Add($"Process exited with code {process.ExitCode}");
                        if (!string.IsNullOrEmpty(errors))
                        {
                            Response.Errors.Add($"stderr: {errors}");
                        }
                        Response.Warnings.Add("Warning: these kind of errors result from an error in the targeted script.");
                        return;
                    }

                    // Parse responses
                    string[] lines = output.Split('\n');
                    var responses = new List<JsonNode>();

                    foreach (string line in lines)
                    {
                        if (string.IsNullOrWhiteSpace(line)) continue;

                        try
                        {
                            JsonNode json = JsonNode.Parse(line);
                            if (json == null) continue;

                            // Validate response has matching key or null key (for init errors)
                            // This ensures we only process responses meant for this request
                            var keyItem = json["key"];
                            if (keyItem == null)
                            {
                                // Key is null (for init errors) - accept this response
                                responses.Add(json);
                            }
                            else
                            {
                                try
                                {
                                    string responseKey = keyItem.GetValue<string>();
                                    if (responseKey == _key)
                                    {
                                        responses.Add(json);
                                    }
                                }
                                catch
                                {
                                    // Invalid key format - ignore
                                }
                            }
                        }
                        catch
                        {
                            // Ignore lines that aren't valid JSON (e.g., debug prints)
                        }
                    }

                    // Process responses
                    if (responses.Count > 0)
                    {
                        bool failure = false;

                        foreach (var resp in responses)
                        {
                            var status = resp["request_status"];
                            if (status != null && status.GetValue<bool?>() == false)
                            {
                                failure = true;
                            }

                            var errorsList = resp["errors"]?.AsArray();
                            if (errorsList != null)
                            {
                                foreach (var error in errorsList)
                                {
                                    if (error != null)
                                    {
                                        Response.Errors.Add(error.GetValue<string>());
                                    }
                                }
                            }
                        }

                        Response.RequestStatus = !failure;
                        Response.RequestStatusSet = true;

                        // Collect data
                        if (isUnique && responses.Count == 1)
                        {
                            var respData = responses[0]["data"];
                            if (respData != null)
                            {
                                Response.Data = respData.ToJsonString();
                            }
                        }
                        else if (isUnique && responses.Count > 1)
                        {
                            Response.RequestStatus = false;
                            Response.Data = "";
                            Response.Errors.Add($"Error: Expected 1 output (isUnique=True) but received {responses.Count}.");
                        }
                        else
                        {
                            var dataArray = new JsonArray();
                            foreach (var resp in responses)
                            {
                                var respData = resp["data"];
                                if (respData != null)
                                {
                                    dataArray.Add(JsonNode.Parse(respData.ToJsonString()));
                                }
                            }
                            Response.Data = dataArray.ToJsonString();
                        }
                    }
                    else if (optionalOutput)
                    {
                        Response.RequestStatusSet = false;
                        Response.Warnings.Add("Warning: the output setting is set to optional, and the targeted program didn't gave any output.");
                    }
                    else
                    {
                        Response.RequestStatus = false;
                        Response.RequestStatusSet = true;
                        Response.Errors.Add("Error: OutputManager might not be used or not correctly.");
                    }
                }
            }
            catch (Exception e)
            {
                Response.RequestStatus = false;
                Response.RequestStatusSet = true;
                Response.OptionalOutput = optionalOutput;
                Response.IsUnique = isUnique;
                Response.Warnings.Add("Warning: targeted file not found or can't be executed, consider checking file informations and language dependencies.");
                Response.Errors.Add($"Error: {e.Message}");
            }
        }

        /// <summary>
        /// Get the full response object
        /// </summary>
        /// <returns>The complete response with status, data, errors, warnings.</returns>
        public InputManagerResponse GetResponse()
        {
            return Response;
        }

        /// <summary>
        /// Get the response data if request was successful
        /// </summary>
        /// <returns>The data from the response as JSON string, or empty string if request failed.
        /// Parse the JSON string to get the actual typed value.</returns>
        public string GetData()
        {
            if (Response.RequestStatusSet && Response.RequestStatus)
            {
                return Response.Data;
            }
            return "";
        }
    }

    /// <summary>
    /// OutputManager - Manages receiving requests from other processes and sending responses
    ///
    /// This is a static class - all methods are static.
    /// Must call Init() before using.
    ///
    /// Static Properties:
    ///     Data: The request data as JSON string (accessible after Init())
    ///
    /// Static Methods:
    ///     Init(): Initialize and read request from stdin
    ///     GetData(): Get the request data as JSON string
    ///     GetInt(): Get the request data as integer
    ///     GetDouble(): Get the request data as double
    ///     GetString(): Get the request data as string
    ///     GetBool(): Get the request data as boolean
    ///     Bundle(): Convert any value to JSON string
    ///     Output(data): Send response back via stdout
    ///     Cleanup(): Clean up resources
    /// </summary>
    public static class OutputManager
    {
        private static TextWriter _originalStdout;
        private static string _requestJson;
        private static string _key;
        private static bool _optionalOutput;
        private static bool _isUnique;
        private static bool _requestStatusSet;
        private static bool _requestStatus;
        private static bool _uniqueStateSet;
        private static bool _uniqueState;
        private static bool _initError;
        private static List<string> _errors = new List<string>();
        private static List<string> _warnings = new List<string>();

        public static string Data { get; set; }

        /// <summary>
        /// Initialize OutputManager and read request from stdin
        ///
        /// Must be called before using Output() or accessing Data.
        /// Suppresses stdout to prevent pollution of JSON protocol.
        /// </summary>
        public static void Init()
        {
            // Save original stdout so we can restore it later
            _originalStdout = Console.Out;

            // Redirect stdout to TextWriter.Null to suppress all Console.WriteLine statements
            Console.SetOut(TextWriter.Null);

            // Read the entire stdin (the JSON request from InputManager)
            using (StreamReader reader = new StreamReader(Console.OpenStandardInput()))
            {
                _requestJson = reader.ReadToEnd();
            }

            // Parse JSON
            try
            {
                JsonNode json = JsonNode.Parse(_requestJson);
                if (json != null)
                {
                    var keyItem = json["key"];
                    if (keyItem != null) _key = keyItem.GetValue<string>();

                    var dataItem = json["data"];
                    if (dataItem != null)
                    {
                        Data = dataItem.ToJsonString();
                    }

                    var optItem = json["optionalOutput"];
                    if (optItem != null) _optionalOutput = optItem.GetValue<bool>();

                    var uniqItem = json["isUnique"];
                    if (uniqItem != null) _isUnique = uniqItem.GetValue<bool>();
                }
            }
            catch
            {
                // Parsing error
            }

            // Reset state for new request
            _errors.Clear();
            _warnings.Clear();
            _initError = false;
            _requestStatusSet = false;
            _uniqueStateSet = false;
        }

        /// <summary>
        /// Get the request data as JSON string
        /// </summary>
        /// <returns>Request data as JSON string (parse to get typed value)</returns>
        public static string GetData()
        {
            return Data;
        }

        /// <summary>
        /// Get the request data as integer
        /// </summary>
        /// <returns>The data as int (0 if not a number)</returns>
        public static int GetInt()
        {
            if (string.IsNullOrEmpty(Data)) return 0;
            try
            {
                var node = JsonNode.Parse(Data);
                if (node != null)
                {
                    return node.GetValue<int>();
                }
            }
            catch { }
            return 0;
        }

        /// <summary>
        /// Get the request data as double
        /// </summary>
        /// <returns>The data as double (0.0 if not a number)</returns>
        public static double GetDouble()
        {
            if (string.IsNullOrEmpty(Data)) return 0.0;
            try
            {
                var node = JsonNode.Parse(Data);
                if (node != null)
                {
                    return node.GetValue<double>();
                }
            }
            catch { }
            return 0.0;
        }

        /// <summary>
        /// Get the request data as string
        /// </summary>
        /// <returns>The data as string (empty string if not a string)</returns>
        public static string GetString()
        {
            if (string.IsNullOrEmpty(Data)) return "";
            try
            {
                var node = JsonNode.Parse(Data);
                if (node != null)
                {
                    return node.GetValue<string>();
                }
            }
            catch { }
            return "";
        }

        /// <summary>
        /// Get the request data as boolean
        /// </summary>
        /// <returns>The data as bool (false if not a boolean)</returns>
        public static bool GetBool()
        {
            if (string.IsNullOrEmpty(Data)) return false;
            try
            {
                var node = JsonNode.Parse(Data);
                if (node != null)
                {
                    return node.GetValue<bool>();
                }
            }
            catch { }
            return false;
        }

        /// <summary>
        /// Bundle any value to JSON string for use with Output()
        /// </summary>
        /// <param name="value">Any JSON-serializable value</param>
        /// <returns>JSON string representation</returns>
        public static string Bundle(object value)
        {
            return JsonSerializer.Serialize(value);
        }

        /// <summary>
        /// Send response back to the calling process
        /// </summary>
        /// <param name="data">Data to send as JSON string</param>
        ///
        /// Note:
        ///     Can be called multiple times if isUnique=false in request.
        ///     Will error if called multiple times when isUnique=true.
        public static void Output(string data)
        {
            // Check if OutputManager was initialized
            if (string.IsNullOrEmpty(Data))
            {
                if (!_initError)
                {
                    // Restore original stdout to actually write the response
                    if (_originalStdout != null)
                    {
                        Console.SetOut(_originalStdout);
                    }

                    _requestStatus = false;
                    _errors.Add("Error: OutputManager isn't initialized.");

                    // Build and write JSON response
                    var response = new JsonObject
                    {
                        ["key"] = null,
                        ["request_status"] = false,
                        ["data"] = null,
                        ["optionalOutput"] = _optionalOutput,
                        ["isUnique"] = null,
                        ["errors"] = new JsonArray(_errors.Select(e => JsonValue.Create(e)).ToArray()),
                        ["warnings"] = new JsonArray()
                    };

                    Console.WriteLine(response.ToJsonString());
                    Console.Out.Flush();

                    _initError = true;
                }
                return;
            }

            // Check if we can output based on isUnique setting
            // uniqueStateSet tracks if we've already output once
            if (!_uniqueStateSet || !_isUnique)
            {
                _requestStatus = true;

                // Restore original stdout to actually write the response
                if (_originalStdout != null)
                {
                    Console.SetOut(_originalStdout);
                }

                // Build and write JSON response
                var response = new JsonObject
                {
                    ["key"] = _key,
                    ["request_status"] = true,
                    ["optionalOutput"] = _optionalOutput,
                    ["isUnique"] = _isUnique,
                    ["errors"] = new JsonArray(),
                    ["warnings"] = new JsonArray()
                };

                try
                {
                    response["data"] = JsonNode.Parse(data);
                }
                catch
                {
                    response["data"] = null;
                }

                Console.WriteLine(response.ToJsonString());
                Console.Out.Flush();
            }
            else
            {
                // Multiple outputs when isUnique=true is an error
                _requestStatus = false;
                _errors.Add($"Error: outputs out of bound (isUnique: {_uniqueState}).");

                // Restore original stdout
                if (_originalStdout != null)
                {
                    Console.SetOut(_originalStdout);
                }

                var response = new JsonObject
                {
                    ["key"] = _key,
                    ["request_status"] = false,
                    ["optionalOutput"] = _optionalOutput,
                    ["isUnique"] = _isUnique,
                    ["errors"] = new JsonArray(_errors.Select(e => JsonValue.Create(e)).ToArray()),
                    ["warnings"] = new JsonArray()
                };

                try
                {
                    response["data"] = JsonNode.Parse(data);
                }
                catch
                {
                    response["data"] = null;
                }

                Console.WriteLine(response.ToJsonString());
                Console.Out.Flush();
            }

            // Mark that we've output once
            _uniqueState = _isUnique;
            _uniqueStateSet = true;

            // Re-suppress stdout after writing response
            Console.SetOut(TextWriter.Null);
        }

        /// <summary>
        /// Clean up OutputManager resources
        /// </summary>
        public static void Cleanup()
        {
            _errors.Clear();
            _warnings.Clear();
        }
    }
}